File: test/perf/bunnymark/bunnymark.js
// -----------------------------------------------
// PDG Bunnymark Performance Test
//
// Classic sprite rendering benchmark that measures
// how many sprites can be rendered while maintaining
// 60 FPS (16.67ms per frame)
//
// Based on the standard Bunnymark test used across
// game engines (Unity, Godot, Phaser, etc.)
//
// -----------------------------------------------

console.log("=== PDG BUNNYMARK PERFORMANCE TEST ===");
console.log("Measuring sprite rendering throughput...\n");

// Configuration
var AUTO_ADD_BUNNIES = true;  // Auto-add until FPS drops
var BUNNIES_PER_BATCH = 100;  // How many to add at once
var TARGET_FPS = 60;
var MIN_FPS = 30;
var FRAME_TIME_TARGET_MS = 1000 / TARGET_FPS;  // 16.67ms
var FRAME_TIME_MIN_MS = 1000 / MIN_FPS;  // 33.33ms
var STATS_UPDATE_INTERVAL = 500;  // Update stats every 500ms
var AUTO_ADD_DELAY = 2000;  // Wait 2 seconds between auto-adds

// Check for command line options
for (var i = 0; i < process.argv.length; i++) {
    if (process.argv[i] === '--manual') {
        AUTO_ADD_BUNNIES = false;
        console.log("Manual mode: Press SPACE to add bunnies");
    } else if (process.argv[i] === '--batch' && i + 1 < process.argv.length) {
        BUNNIES_PER_BATCH = parseInt(process.argv[i + 1]);
        console.log("Batch size set to:", BUNNIES_PER_BATCH);
    }
}

if (AUTO_ADD_BUNNIES) {
    console.log("Auto mode: Will add " + BUNNIES_PER_BATCH + " bunnies every " + (AUTO_ADD_DELAY/1000) + " seconds until FPS drops below " + TARGET_FPS);
} else {
    console.log("Manual mode: Press SPACE to add " + BUNNIES_PER_BATCH + " bunnies");
}
console.log("Press ESC to quit and show final results\n");

// Bunny state
var bunnies = [];
var bunnyImage = null;
var port = null;

// Physics constants
var GRAVITY = 0.5;
var BOUNCE_DAMPING = 0.95;

// Stats tracking
var frameCount = 0;
var totalFrameTime = 0;
var maxFrameTime = 0;
var minFrameTime = 999999;
var droppedFrames = 0;
var lastStatsUpdate = 0;
var currentFPS = 60;
var currentFrameTimeMs = 0;
var frameTimes = [];  // For calculating percentiles
var maxBunniesAt60FPS = 0;
var maxBunniesAt30FPS = 0;
var testStartTime = 0;
var lastAutoAddTime = 0;

// Performance metrics
var metrics = {
    totalFrames: 0,
    meanFrameTime: 0,
    minFrameTime: 999999,
    maxFrameTime: 0,
    percentile95: 0,
    percentile99: 0,
    droppedFrames: 0,
    maxBunniesAt60FPS: 0,
    maxBunniesAt30FPS: 0,
    testDurationSeconds: 0
};

// Bunny constructor
function Bunny(image, bounds) {
    this.x = Math.random() * bounds.width();
    this.y = Math.random() * bounds.height();
    this.speedX = (Math.random() - 0.5) * 10;
    this.speedY = (Math.random() - 0.5) * 10 - 5; // Bias upward
    this.image = image;
}

Bunny.prototype.update = function(bounds) {
    // Apply gravity
    this.speedY += GRAVITY;
    
    // Update position
    this.x += this.speedX;
    this.y += this.speedY;
    
    // Bounce off walls
    var imgWidth = this.image.getWidth();
    var imgHeight = this.image.getHeight();
    
    if (this.x <= 0) {
        this.x = 0;
        this.speedX = Math.abs(this.speedX) * BOUNCE_DAMPING;
    } else if (this.x + imgWidth >= bounds.width()) {
        this.x = bounds.width() - imgWidth;
        this.speedX = -Math.abs(this.speedX) * BOUNCE_DAMPING;
    }
    
    if (this.y <= 0) {
        this.y = 0;
        this.speedY = Math.abs(this.speedY) * BOUNCE_DAMPING;
    } else if (this.y + imgHeight >= bounds.height()) {
        this.y = bounds.height() - imgHeight;
        this.speedY = -Math.abs(this.speedY) * BOUNCE_DAMPING;
    }
};

Bunny.prototype.draw = function(port) {
    var pos = new pdg.Point(this.x, this.y);
    port.drawImage(this.image, pos, new pdg.Attributes());
};

// Add bunnies
function addBunnies(count) {
    var bounds = port.getDrawingArea();
    for (var i = 0; i < count; i++) {
        bunnies.push(new Bunny(bunnyImage, bounds));
    }
    console.log("Added " + count + " bunnies. Total: " + bunnies.length);
}

// Calculate percentile
function calculatePercentile(sortedArray, percentile) {
    if (sortedArray.length === 0) return 0;
    var index = Math.ceil((percentile / 100) * sortedArray.length) - 1;
    index = Math.max(0, Math.min(index, sortedArray.length - 1));
    return sortedArray[index];
}

// Update statistics
function updateStats(frameTimeMs) {
    frameCount++;
    totalFrameTime += frameTimeMs;
    maxFrameTime = Math.max(maxFrameTime, frameTimeMs);
    minFrameTime = Math.min(minFrameTime, frameTimeMs);
    
    // Track frame time for percentile calculation
    frameTimes.push(frameTimeMs);
    
    // Count dropped frames (> 16.67ms = below 60 FPS)
    if (frameTimeMs > FRAME_TIME_TARGET_MS) {
        droppedFrames++;
    }
    
    // Track max bunnies at different FPS thresholds
    if (frameTimeMs <= FRAME_TIME_TARGET_MS && bunnies.length > maxBunniesAt60FPS) {
        maxBunniesAt60FPS = bunnies.length;
    }
    if (frameTimeMs <= FRAME_TIME_MIN_MS && bunnies.length > maxBunniesAt30FPS) {
        maxBunniesAt30FPS = bunnies.length;
    }
}

// Print final results
function printFinalResults() {
    console.log("\n" + "=".repeat(60));
    console.log("BUNNYMARK TEST RESULTS");
    console.log("=".repeat(60));
    
    // Calculate final metrics
    var testDuration = (Date.now() - testStartTime) / 1000;
    var meanFrameTime = totalFrameTime / frameCount;
    
    // Sort frame times for percentiles
    frameTimes.sort(function(a, b) { return a - b; });
    var p95 = calculatePercentile(frameTimes, 95);
    var p99 = calculatePercentile(frameTimes, 99);
    
    console.log("\nTest Duration: " + testDuration.toFixed(2) + " seconds");
    console.log("Total Frames: " + frameCount);
    console.log("\nFRAME TIME STATISTICS:");
    console.log("  Mean:        " + meanFrameTime.toFixed(2) + " ms");
    console.log("  Min:         " + minFrameTime.toFixed(2) + " ms");
    console.log("  Max:         " + maxFrameTime.toFixed(2) + " ms");
    console.log("  95th %ile:   " + p95.toFixed(2) + " ms");
    console.log("  99th %ile:   " + p99.toFixed(2) + " ms");
    console.log("\nPERFORMANCE METRICS:");
    console.log("  Average FPS:           " + (1000 / meanFrameTime).toFixed(2));
    console.log("  Dropped Frames:        " + droppedFrames + " (" + ((droppedFrames / frameCount) * 100).toFixed(1) + "%)");
    console.log("  Max Bunnies @ 60 FPS:  " + maxBunniesAt60FPS);
    console.log("  Max Bunnies @ 30 FPS:  " + maxBunniesAt30FPS);
    console.log("  Final Bunny Count:     " + bunnies.length);
    
    console.log("\nKEY METRIC (Bunnymark Score): " + maxBunniesAt60FPS + " sprites @ 60 FPS");
    console.log("=".repeat(60) + "\n");
    
    // Export results to JSON
    var results = {
        testName: "PDG Bunnymark",
        timestamp: new Date().toISOString(),
        testDurationSeconds: testDuration,
        totalFrames: frameCount,
        frameTime: {
            meanMs: parseFloat(meanFrameTime.toFixed(2)),
            minMs: parseFloat(minFrameTime.toFixed(2)),
            maxMs: parseFloat(maxFrameTime.toFixed(2)),
            percentile95Ms: parseFloat(p95.toFixed(2)),
            percentile99Ms: parseFloat(p99.toFixed(2))
        },
        performance: {
            averageFPS: parseFloat((1000 / meanFrameTime).toFixed(2)),
            droppedFrames: droppedFrames,
            droppedFramePercent: parseFloat(((droppedFrames / frameCount) * 100).toFixed(1)),
            maxBunniesAt60FPS: maxBunniesAt60FPS,
            maxBunniesAt30FPS: maxBunniesAt30FPS,
            finalBunnyCount: bunnies.length
        },
        bunnymarkScore: maxBunniesAt60FPS
    };
    
    // Save results
    var fs = require('fs');
    var resultsFile = 'test/perf/bunnymark/bunnymark_results.json';
    try {
        fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2));
        console.log("Results saved to: " + resultsFile);
    } catch (e) {
        console.log("Warning: Could not save results to file:", e.message);
    }
}

// Setup graphics
function setup() {
    console.log("Setting up graphics...");
    
    // Create window
    var windowRect = new pdg.Rect(800, 600);
    port = pdg.gfx.createWindowPort(windowRect, "PDG Bunnymark Test");
    
    if (!port) {
        console.log("ERROR: Could not create window port");
        process.exit(1);
    }
    
    console.log("Window created: " + port.getDrawingArea().width() + "x" + port.getDrawingArea().height());
    
    // Load bunny image - try multiple paths
    // Using the classic Bunnymark wabbit.png sprite (26x37)
    var imagePaths = [
        "test/perf/bunnymark/wabbit.png",
        "perf/bunnymark/wabbit.png",
        "bunnymark/wabbit.png",
        "wabbit.png"
    ];
    
    for (var i = 0; i < imagePaths.length; i++) {
        try {
            bunnyImage = new pdg.Image(imagePaths[i]);
            console.log("Loaded bunny image from: " + imagePaths[i]);
            console.log("Image size: " + bunnyImage.getWidth() + "x" + bunnyImage.getHeight());
            break;
        } catch (e) {
            // Try next path
        }
    }
    
    if (!bunnyImage) {
        console.log("ERROR: Could not load bunny image from any path");
        process.exit(1);
    }
    
    // Start with initial batch of bunnies
    addBunnies(BUNNIES_PER_BATCH);
    
    testStartTime = Date.now();
    lastAutoAddTime = testStartTime;
    lastStatsUpdate = testStartTime;
}

// Draw handler
var lastFrameTime = Date.now();

function onDraw(evt) {
    if (evt.port !== port) return false;
    
    var frameStartTime = Date.now();
    var deltaTime = frameStartTime - lastFrameTime;
    lastFrameTime = frameStartTime;
    
    // Clear background
    var bounds = port.getDrawingArea();
    var bgAttrs = new pdg.Attributes().fillColor(new pdg.Color(0.1, 0.1, 0.15, 1.0));
    port.drawRect(bounds, bgAttrs);
    
    // Update and draw all bunnies
    for (var i = 0; i < bunnies.length; i++) {
        bunnies[i].update(bounds);
        bunnies[i].draw(port);
    }
    
    // Draw stats overlay
    var textAttrs = new pdg.Attributes()
        .textSize(16)
        .textStyle(pdg.textStyle_Plain)
        .fillColor("white");
    
    var y = 20;
    var lineHeight = 20;
    
    port.drawText("PDG BUNNYMARK TEST", new pdg.Point(10, y), textAttrs);
    y += lineHeight * 1.5;
    
    port.drawText("Bunnies: " + bunnies.length, new pdg.Point(10, y), textAttrs);
    y += lineHeight;
    
    port.drawText("FPS: " + currentFPS.toFixed(1), new pdg.Point(10, y), textAttrs);
    y += lineHeight;
    
    port.drawText("Frame Time: " + currentFrameTimeMs.toFixed(2) + " ms", new pdg.Point(10, y), textAttrs);
    y += lineHeight;
    
    port.drawText("Max @ 60 FPS: " + maxBunniesAt60FPS, new pdg.Point(10, y), textAttrs);
    y += lineHeight;
    
    port.drawText("Max @ 30 FPS: " + maxBunniesAt30FPS, new pdg.Point(10, y), textAttrs);
    y += lineHeight * 1.5;
    
    if (AUTO_ADD_BUNNIES) {
        port.drawText("Auto-adding bunnies...", new pdg.Point(10, y), textAttrs);
    } else {
        port.drawText("Press SPACE to add bunnies", new pdg.Point(10, y), textAttrs);
    }
    y += lineHeight;
    
    port.drawText("Press ESC to quit", new pdg.Point(10, y), textAttrs);
    
    // Calculate frame time
    var frameEndTime = Date.now();
    var frameTimeMs = frameEndTime - frameStartTime;
    
    // Update stats
    updateStats(frameTimeMs);
    currentFrameTimeMs = frameTimeMs;
    
    // Update current FPS periodically
    if (frameEndTime - lastStatsUpdate > STATS_UPDATE_INTERVAL) {
        var recentFrames = frameCount - (frameCount - 30);  // Rough approximation
        if (frameTimeMs > 0) {
            currentFPS = 1000 / frameTimeMs;
        }
        lastStatsUpdate = frameEndTime;
    }
    
    // Auto-add bunnies if enabled and enough time has passed
    if (AUTO_ADD_BUNNIES && (frameEndTime - lastAutoAddTime) > AUTO_ADD_DELAY) {
        // Only add more if we're still above minimum FPS
        if (currentFrameTimeMs < FRAME_TIME_MIN_MS) {
            addBunnies(BUNNIES_PER_BATCH);
            lastAutoAddTime = frameEndTime;
        } else {
            console.log("Stopped auto-adding: FPS dropped below " + MIN_FPS);
            console.log("Run test for a bit longer, then press ESC to see final results");
            AUTO_ADD_BUNNIES = false;  // Stop auto-adding
        }
    }
    
    return true;
}

// Key handler
function onKeyPress(evt) {
    if (evt.unicode == pdg.key_Escape) {
        console.log("\nTest stopped by user.");
        printFinalResults();
        pdg.gfx.closeGraphicsPort(port);
        pdg.quit();
        return true;
    } else if (evt.unicode == 32) {  // Space key
        addBunnies(BUNNIES_PER_BATCH);
        lastAutoAddTime = Date.now();  // Reset auto-add timer
        return true;
    }
    return false;
}

// Setup and run
setup();

pdg.on(pdg.eventType_PortDraw, onDraw);
pdg.on(pdg.eventType_KeyPress, onKeyPress);

console.log("\nTest started. Watch the window for live stats.");
console.log("Starting with " + bunnies.length + " bunnies...\n");

pdg.run();

File: test/perf/cpp-bunnymark/bunnymark.cpp
// -----------------------------------------------
// bunnymark.cpp
//
// PDG Bunnymark Performance Test (C++ version)
//
// Classic sprite rendering benchmark that measures
// how many sprites can be rendered while maintaining
// 60 FPS (16.67ms per frame)
//
// -----------------------------------------------

#include "pdg_project.h"
#include "pdg/framework.h"
#include <vector>
#include <string>
#include <sstream>
#include <fstream>
#include <algorithm>
#include <cmath>
#include <cstdlib>

using namespace pdg;

// Configuration
const int BUNNIES_PER_BATCH = 100;
const int TARGET_FPS = 60;
const int MIN_FPS = 30;
const float FRAME_TIME_TARGET_MS = 1000.0f / TARGET_FPS;  // 16.67ms
const float FRAME_TIME_MIN_MS = 1000.0f / MIN_FPS;        // 33.33ms
const int STATS_UPDATE_INTERVAL = 500;  // Update stats every 500ms
const int AUTO_ADD_DELAY = 2000;        // Wait 2 seconds between auto-adds

// Physics constants
const float GRAVITY = 0.5f;
const float BOUNCE_DAMPING = 0.95f;

// Window size
const int WINDOW_WIDTH = 800;
const int WINDOW_HEIGHT = 600;

// Bunny structure
struct Bunny {
    float x;
    float y;
    float speedX;
    float speedY;
    
    Bunny(float initX, float initY, float spdX, float spdY)
        : x(initX), y(initY), speedX(spdX), speedY(spdY) {}
};

// Global state
Port* gPort = nullptr;
Image* gBunnyImage = nullptr;
std::vector<Bunny> gBunnies;

// Stats tracking
int gFrameCount = 0;
float gTotalFrameTime = 0;
float gMaxFrameTime = 0;
float gMinFrameTime = 999999.0f;
int gDroppedFrames = 0;
ms_time gLastStatsUpdate = 0;
float gCurrentFPS = 60.0f;
float gCurrentFrameTimeMs = 0;
std::vector<float> gFrameTimes;
int gMaxBunniesAt60FPS = 0;
int gMaxBunniesAt30FPS = 0;
ms_time gTestStartTime = 0;
ms_time gLastAutoAddTime = 0;
ms_time gLastFrameTime = 0;
bool gAutoAddBunnies = true;
bool gTestRunning = true;

// Function declarations
void addBunnies(int count);
void updateStats(float frameTimeMs);
void printFinalResults();
void saveResultsJSON(const char* filename);
float calculatePercentile(std::vector<float>& sortedArray, float percentile);

// Random float between 0 and 1
inline float randomFloat() {
    return static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
}

// Add bunnies to the scene
void addBunnies(int count) {
    for (int i = 0; i < count; ++i) {
        float x = randomFloat() * WINDOW_WIDTH;
        float y = randomFloat() * WINDOW_HEIGHT;
        float speedX = (randomFloat() - 0.5f) * 10.0f;
        float speedY = (randomFloat() - 0.5f) * 10.0f - 5.0f; // Bias upward
        gBunnies.push_back(Bunny(x, y, speedX, speedY));
    }
    std::cout << "Added " << count << " bunnies. Total: " << gBunnies.size() << std::endl;
}

// Update statistics
void updateStats(float frameTimeMs) {
    gFrameCount++;
    gTotalFrameTime += frameTimeMs;
    gMaxFrameTime = std::max(gMaxFrameTime, frameTimeMs);
    gMinFrameTime = std::min(gMinFrameTime, frameTimeMs);
    
    // Track frame time for percentile calculation
    gFrameTimes.push_back(frameTimeMs);
    
    // Count dropped frames (> 16.67ms = below 60 FPS)
    if (frameTimeMs > FRAME_TIME_TARGET_MS) {
        gDroppedFrames++;
    }
    
    // Track max bunnies at different FPS thresholds
    if (frameTimeMs <= FRAME_TIME_TARGET_MS && (int)gBunnies.size() > gMaxBunniesAt60FPS) {
        gMaxBunniesAt60FPS = gBunnies.size();
    }
    if (frameTimeMs <= FRAME_TIME_MIN_MS && (int)gBunnies.size() > gMaxBunniesAt30FPS) {
        gMaxBunniesAt30FPS = gBunnies.size();
    }
}

// Calculate percentile from sorted array
float calculatePercentile(std::vector<float>& sortedArray, float percentile) {
    if (sortedArray.empty()) return 0;
    size_t index = static_cast<size_t>(std::ceil((percentile / 100.0f) * sortedArray.size())) - 1;
    index = std::min(index, sortedArray.size() - 1);
    return sortedArray[index];
}

// Print final results to console
void printFinalResults() {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "BUNNYMARK TEST RESULTS (C++)" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    
    // Calculate final metrics
    float testDuration = (OS::getMilliseconds() - gTestStartTime) / 1000.0f;
    float meanFrameTime = gTotalFrameTime / gFrameCount;
    
    // Sort frame times for percentiles
    std::sort(gFrameTimes.begin(), gFrameTimes.end());
    float p95 = calculatePercentile(gFrameTimes, 95.0f);
    float p99 = calculatePercentile(gFrameTimes, 99.0f);
    
    std::cout << "\nTest Duration: " << testDuration << " seconds" << std::endl;
    std::cout << "Total Frames: " << gFrameCount << std::endl;
    std::cout << "\nFRAME TIME STATISTICS:" << std::endl;
    std::cout << "  Mean:        " << meanFrameTime << " ms" << std::endl;
    std::cout << "  Min:         " << gMinFrameTime << " ms" << std::endl;
    std::cout << "  Max:         " << gMaxFrameTime << " ms" << std::endl;
    std::cout << "  95th %ile:   " << p95 << " ms" << std::endl;
    std::cout << "  99th %ile:   " << p99 << " ms" << std::endl;
    std::cout << "\nPERFORMANCE METRICS:" << std::endl;
    std::cout << "  Average FPS:           " << (1000.0f / meanFrameTime) << std::endl;
    std::cout << "  Dropped Frames:        " << gDroppedFrames 
              << " (" << ((float)gDroppedFrames / gFrameCount * 100.0f) << "%)" << std::endl;
    std::cout << "  Max Bunnies @ 60 FPS:  " << gMaxBunniesAt60FPS << std::endl;
    std::cout << "  Max Bunnies @ 30 FPS:  " << gMaxBunniesAt30FPS << std::endl;
    std::cout << "  Final Bunny Count:     " << gBunnies.size() << std::endl;
    
    std::cout << "\nKEY METRIC (Bunnymark Score): " << gMaxBunniesAt60FPS << " sprites @ 60 FPS" << std::endl;
    std::cout << std::string(60, '=') << "\n" << std::endl;
}

// Save results as JSON
void saveResultsJSON(const char* filename) {
    float testDuration = (OS::getMilliseconds() - gTestStartTime) / 1000.0f;
    float meanFrameTime = gTotalFrameTime / gFrameCount;
    
    // Sort frame times for percentiles
    std::sort(gFrameTimes.begin(), gFrameTimes.end());
    float p95 = calculatePercentile(gFrameTimes, 95.0f);
    float p99 = calculatePercentile(gFrameTimes, 99.0f);
    
    // Get current timestamp (simplified)
    time_t now = time(0);
    char timestamp[80];
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%dT%H:%M:%S", gmtime(&now));
    
    std::ofstream outfile(filename);
    if (!outfile.is_open()) {
        std::cerr << "Warning: Could not save results to file: " << filename << std::endl;
        return;
    }
    
    outfile << "{\n";
    outfile << "  \"testName\": \"PDG Bunnymark (C++)\",\n";
    outfile << "  \"timestamp\": \"" << timestamp << "Z\",\n";
    outfile << "  \"testDurationSeconds\": " << testDuration << ",\n";
    outfile << "  \"totalFrames\": " << gFrameCount << ",\n";
    outfile << "  \"frameTime\": {\n";
    outfile << "    \"meanMs\": " << meanFrameTime << ",\n";
    outfile << "    \"minMs\": " << gMinFrameTime << ",\n";
    outfile << "    \"maxMs\": " << gMaxFrameTime << ",\n";
    outfile << "    \"percentile95Ms\": " << p95 << ",\n";
    outfile << "    \"percentile99Ms\": " << p99 << "\n";
    outfile << "  },\n";
    outfile << "  \"performance\": {\n";
    outfile << "    \"averageFPS\": " << (1000.0f / meanFrameTime) << ",\n";
    outfile << "    \"droppedFrames\": " << gDroppedFrames << ",\n";
    outfile << "    \"droppedFramePercent\": " << ((float)gDroppedFrames / gFrameCount * 100.0f) << ",\n";
    outfile << "    \"maxBunniesAt60FPS\": " << gMaxBunniesAt60FPS << ",\n";
    outfile << "    \"maxBunniesAt30FPS\": " << gMaxBunniesAt30FPS << ",\n";
    outfile << "    \"finalBunnyCount\": " << gBunnies.size() << "\n";
    outfile << "  },\n";
    outfile << "  \"bunnymarkScore\": " << gMaxBunniesAt60FPS << "\n";
    outfile << "}\n";
    
    outfile.close();
    std::cout << "Results saved to: " << filename << std::endl;
}

// Event handler class
class BunnymarkHandler : public IEventHandler {
public:
    virtual bool handleEvent(EventEmitter* inEmitter, long inEventType, void* inEventData) throw() {
        if (inEventType == eventType_PortDraw) {
            if (!gTestRunning) return false;
            
            PortDrawInfo* drawInfo = static_cast<PortDrawInfo*>(inEventData);
            Port* port = drawInfo->port;
            
            ms_time frameStartTime = OS::getMilliseconds();
            gLastFrameTime = frameStartTime;
            
            // Clear background
            pdg::Rect bounds = port->getDrawingArea();
            pdg::Attributes bgAttrs;
            bgAttrs.fillColor(pdg::Color(0.1f, 0.1f, 0.15f, 1.0f));
            port->drawRect(bounds, bgAttrs);
            
            // Get image dimensions once
            float imgWidth = gBunnyImage->getWidth();
            float imgHeight = gBunnyImage->getHeight();
            
            // Update and draw all bunnies
            for (size_t i = 0; i < gBunnies.size(); ++i) {
                Bunny& bunny = gBunnies[i];
                
                // Apply gravity
                bunny.speedY += GRAVITY;
                
                // Update position
                bunny.x += bunny.speedX;
                bunny.y += bunny.speedY;
                
                // Bounce off walls
                if (bunny.x <= 0) {
                    bunny.x = 0;
                    bunny.speedX = std::abs(bunny.speedX) * BOUNCE_DAMPING;
                } else if (bunny.x + imgWidth >= bounds.width()) {
                    bunny.x = bounds.width() - imgWidth;
                    bunny.speedX = -std::abs(bunny.speedX) * BOUNCE_DAMPING;
                }
                
                if (bunny.y <= 0) {
                    bunny.y = 0;
                    bunny.speedY = std::abs(bunny.speedY) * BOUNCE_DAMPING;
                } else if (bunny.y + imgHeight >= bounds.height()) {
                    bunny.y = bounds.height() - imgHeight;
                    bunny.speedY = -std::abs(bunny.speedY) * BOUNCE_DAMPING;
                }
                
                // Draw bunny
                pdg::Point pos(bunny.x, bunny.y);
                pdg::Attributes attrs;
                port->drawImage(gBunnyImage, pos, attrs);
            }
            
            // Draw stats overlay
            pdg::Attributes textAttrs;
            textAttrs.textSize(16);
            textAttrs.textStyle(pdg::textStyle_Plain);
            textAttrs.fillColor(pdg::Color(1.0f, 1.0f, 1.0f, 1.0f));
            
            float y = 20;
            float lineHeight = 20;
            
            port->drawText("PDG BUNNYMARK TEST (C++)", pdg::Point(10, y), textAttrs);
            y += lineHeight * 1.5f;
            
            char buffer[256];
            
            snprintf(buffer, sizeof(buffer), "Bunnies: %zu", gBunnies.size());
            port->drawText(buffer, pdg::Point(10, y), textAttrs);
            y += lineHeight;
            
            snprintf(buffer, sizeof(buffer), "FPS: %.1f", gCurrentFPS);
            port->drawText(buffer, pdg::Point(10, y), textAttrs);
            y += lineHeight;
            
            snprintf(buffer, sizeof(buffer), "Frame Time: %.2f ms", gCurrentFrameTimeMs);
            port->drawText(buffer, pdg::Point(10, y), textAttrs);
            y += lineHeight;
            
            snprintf(buffer, sizeof(buffer), "Max @ 60 FPS: %d", gMaxBunniesAt60FPS);
            port->drawText(buffer, pdg::Point(10, y), textAttrs);
            y += lineHeight;
            
            snprintf(buffer, sizeof(buffer), "Max @ 30 FPS: %d", gMaxBunniesAt30FPS);
            port->drawText(buffer, pdg::Point(10, y), textAttrs);
            y += lineHeight * 1.5f;
            
            if (gAutoAddBunnies) {
                port->drawText("Auto-adding bunnies...", pdg::Point(10, y), textAttrs);
            } else {
                port->drawText("Press SPACE to add bunnies", pdg::Point(10, y), textAttrs);
            }
            y += lineHeight;
            
            port->drawText("Press ESC to quit", pdg::Point(10, y), textAttrs);
            
            // Calculate frame time
            ms_time frameEndTime = OS::getMilliseconds();
            float frameTimeMs = static_cast<float>(frameEndTime - frameStartTime);
            
            // Update stats
            updateStats(frameTimeMs);
            gCurrentFrameTimeMs = frameTimeMs;
            
            // Update current FPS periodically
            if (frameEndTime - gLastStatsUpdate > STATS_UPDATE_INTERVAL) {
                if (frameTimeMs > 0) {
                    gCurrentFPS = 1000.0f / frameTimeMs;
                }
                gLastStatsUpdate = frameEndTime;
            }
            
            // Auto-add bunnies if enabled and enough time has passed
            if (gAutoAddBunnies && (frameEndTime - gLastAutoAddTime) > AUTO_ADD_DELAY) {
                // Only add more if we're still above minimum FPS
                if (gCurrentFrameTimeMs < FRAME_TIME_MIN_MS) {
                    addBunnies(BUNNIES_PER_BATCH);
                    gLastAutoAddTime = frameEndTime;
                } else {
                    std::cout << "Stopped auto-adding: FPS dropped below " << MIN_FPS << std::endl;
                    std::cout << "Run test for a bit longer, then press ESC to see final results" << std::endl;
                    gAutoAddBunnies = false;  // Stop auto-adding
                }
            }
            
            return true;
        }
        else if (inEventType == eventType_KeyPress) {
            KeyPressInfo* keyInfo = static_cast<KeyPressInfo*>(inEventData);
            if (keyInfo->unicode == key_Escape) {
                std::cout << "\nTest stopped by user." << std::endl;
                printFinalResults();
                saveResultsJSON("test/perf/cpp-bunnymark/bunnymark_results.json");
                gTestRunning = false;
                if (gPort) {
                    GraphicsManager::instance().closeGraphicsPort(gPort);
                }
                OS::exit(0);
                return true;
            } else if (keyInfo->unicode == 32) {  // Space key
                addBunnies(BUNNIES_PER_BATCH);
                gLastAutoAddTime = OS::getMilliseconds();  // Reset auto-add timer
                return true;
            }
        }
        else if (inEventType == eventType_Startup) {
            // Handle startup here
            std::cout << "=== PDG BUNNYMARK PERFORMANCE TEST (C++) ===" << std::endl;
            std::cout << "Measuring sprite rendering throughput...\n" << std::endl;
            std::cout << "Auto mode: Will add " << BUNNIES_PER_BATCH << " bunnies every " 
                      << (AUTO_ADD_DELAY/1000) << " seconds until FPS drops below " << TARGET_FPS << std::endl;
            std::cout << "Press ESC to quit and show final results\n" << std::endl;
            
            // Create window
            pdg::Rect windowRect(WINDOW_WIDTH, WINDOW_HEIGHT);
            gPort = GraphicsManager::instance().createWindowPort(windowRect, "PDG Bunnymark Test (C++)");
            
            if (!gPort) {
                std::cerr << "ERROR: Could not create window port" << std::endl;
                OS::exit(1);
                return true;
            }
            
            std::cout << "Window created: " << gPort->getDrawingArea().width() 
                      << "x" << gPort->getDrawingArea().height() << std::endl;
            
            // Load bunny image - try multiple paths
            const char* imagePaths[] = {
                "test/perf/bunnymark/wabbit.png",
                "perf/bunnymark/wabbit.png",
                "bunnymark/wabbit.png",
                "../bunnymark/wabbit.png",
                "wabbit.png"
            };
            
            for (size_t i = 0; i < sizeof(imagePaths) / sizeof(imagePaths[0]); ++i) {
                try {
                    gBunnyImage = Image::createImageFromFile(imagePaths[i]);
                    if (gBunnyImage) {
                        std::cout << "Loaded bunny image from: " << imagePaths[i] << std::endl;
                        std::cout << "Image size: " << gBunnyImage->getWidth() 
                                  << "x" << gBunnyImage->getHeight() << std::endl;
                        break;
                    }
                } catch (...) {
                    // Try next path
                }
            }
            
            if (!gBunnyImage) {
                std::cerr << "ERROR: Could not load bunny image from any path" << std::endl;
                OS::exit(1);
                return true;
            }
            
            // Initialize random seed
            srand(static_cast<unsigned int>(time(NULL)));
            
            // Start with initial batch of bunnies
            addBunnies(BUNNIES_PER_BATCH);
            
            gTestStartTime = OS::getMilliseconds();
            gLastAutoAddTime = gTestStartTime;
            gLastStatsUpdate = gTestStartTime;
            gLastFrameTime = gTestStartTime;
            
            std::cout << "\nTest started. Watch the window for live stats." << std::endl;
            std::cout << "Starting with " << gBunnies.size() << " bunnies...\n" << std::endl;
            
            return true;
        }
        
        return false;
    }
};

// Initialize PDG framework
namespace pdg {

bool Initializer::allowHorizontalOrientation() throw() { return true; }
bool Initializer::allowVerticalOrientation() throw() { return true; }
const char* Initializer::getAppName(bool haveMainResourceFile) throw() { return "PDG Bunnymark (C++)"; }
const char* Initializer::getMainResourceFileName() throw() { return NULL; }

bool Initializer::installGlobalHandlers() throw() {
    BunnymarkHandler* handler = new BunnymarkHandler();
    EventManager::instance().addHandler(handler);
    return true;
}

bool Initializer::getGraphicsEnvironmentDimensions(Rect maxWindowDim, Rect maxFullScreenDim,
    long& ioWidth, long& ioHeight, uint8& ioDepth) throw()
{
    ioWidth = WINDOW_WIDTH;
    ioHeight = WINDOW_HEIGHT;
    ioDepth = 32;
    return false;  // windowed mode
}

} // end namespace pdg

